pkg/types/validation/installconfig.go

```diff
@@ -90,35 +89,56 @@ func validateNetworking(n *types.Networking, fldPath *field.Path) field.ErrorLis
+			}
+		}
+	}
+
+	if len(n.ServiceNetwork) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("serviceNetwork"), "a service network is required"))
+	}
+
+	for i, cn := range n.ClusterNetwork {
+		allErrs = append(allErrs, validateClusterNetwork(n, &cn, i, fldPath.Child("clusterNetwork").Index(i))...)
 	}
-	if len(n.ClusterNetworks) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("clusterNetworks"), "cluster network required"))
+	if len(n.ClusterNetwork) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterNetwork"), "cluster network required"))
 	}
 	return allErrs
 }
 
-func validateClusterNetwork(cn *types.ClusterNetworkEntry, fldPath *field.Path, serviceCIDR *net.IPNet) field.ErrorList {
+func validateClusterNetwork(n *types.Networking, cn *types.ClusterNetworkEntry, idx int, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 	if err := validate.SubnetCIDR(&cn.CIDR.IPNet); err != nil {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("cidr"), cn.CIDR.IPNet.String(), err.Error()))
 	}
-	if validate.DoCIDRsOverlap(&cn.CIDR.IPNet, serviceCIDR) {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("cidr"), cn.CIDR.String(), "cluster network CIDR must not overlap with serviceCIDR"))
+	if n.MachineCIDR != nil && validate.DoCIDRsOverlap(&cn.CIDR.IPNet, &n.MachineCIDR.IPNet) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("cidr"), cn.CIDR.String(), "cluster network must not overlap with machine CIDR"))
+	}
+	for i, sn := range n.ServiceNetwork {
+		if validate.DoCIDRsOverlap(&cn.CIDR.IPNet, &sn.IPNet) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("cidr"), cn.CIDR.String(), fmt.Sprintf("cluster network must not overlap with service network %d", i)))
+		}
+	}
+	for i, acn := range n.ClusterNetwork[0:idx] {
+		if validate.DoCIDRsOverlap(&cn.CIDR.IPNet, &acn.CIDR.IPNet) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("cidr"), cn.CIDR.String(), fmt.Sprintf("cluster network must not overlap with cluster network %d", i)))
+		}
 	}
-	if cn.HostSubnetLength < 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostSubnetLength"), cn.HostSubnetLength, "hostSubnetLength must be positive"))
+	if cn.HostPrefix < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostPrefix"), cn.HostPrefix, "hostPrefix must be positive"))
 	}
-	if ones, bits := cn.CIDR.Mask.Size(); cn.HostSubnetLength > int32(bits-ones) {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostSubnetLength"), cn.HostSubnetLength, "cluster network host subnet must not be larger than CIDR "+cn.CIDR.String()))
+	if ones, _ := cn.CIDR.Mask.Size(); cn.HostPrefix < int32(ones) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostPrefix"), cn.HostPrefix, "cluster network host prefix must not be larger than CIDR "+cn.CIDR.String()))
```

Comment:
s/larger/smaller/

Reply:
Remember, in networking, 20 is larger than 24 :-)

Reply:
I would agree if the message said "cluster network *size* must not be larger". But the message is talking about the host prefix itself not the network size. Would it make sense for the message to be something like "number of hosts must not be larger than number of addresses"?

Reply:
But if it makes sense in network parlance as is, then I can live with it.

Reply:
yeah, using "size" is a good idea - tweaked the wording a bit.